### PR TITLE
Update for Jlab 3

### PR DESCRIPTION
**Pull request by ianhi**
Building off of https://github.com/jupyter-widgets/widget-ts-cookiecutter/pull/83

I tested this locally and it worked in both classic notebook and jlab3.

Things I did not do:
1. Update python packaging

Waiting until https://github.com/jupyter-widgets/widget-cookiecutter/pull/87 resolves.

2. Fix the tests

They were broken before and there is a draft fix in https://github.com/jupyter-widgets/widget-ts-cookiecutter/pull/86 so I didn't want to overcomplicate this PR.

### CODE REVIEW

**Comment by ianhi** [comment]
attn all those reviewing the previous PR: @martinRenou @jtpio @jasongrout @vidartf

**Comment by jtpio** [comment]
Thanks @ianhi for picking this up!

Changes look good. Maybe it could be worth checking the state of https://github.com/jupyter-widgets/widget-ts-cookiecutter/pull/86 to see if it can already be merged.

**Comment by ianhi** [comment]
Looks like the output of `python -c "from notebook.nbextensions import validate_nbextension; import sys; sys.exit(validate_nbextension('jupyter_widget_testwidgets/extension') or 0)"`


is getting mangled somehow into:
```
['        - require? \x1b[31m X\x1b[0m jupyter_widget_testwidgets/extension']
```

Which I'm also seeing locally. Is that a real problem with the widget or an issue with python and printing out color characters?

**Comment by ianhi** [comment]
Oh hmm, there's clearly an `X` in there, but the `enable` command worked fine gives an `OK`
image

**Comment by ianhi** [comment]
It turns out it is a real error. This is the second time I accidentally broke notebook while working on this PR.

**Comment by jtpio** [comment]
> Which I'm also seeing locally. Is that a real problem with the widget or an issue with python and printing out color characters?

Probably because of the coloring?

What does `jupyter nbextension list` give locally? Could it be an issue with the nbextension data files not being copied to the right location?

**Comment by ianhi** [comment]
Looking into it more I think this is due to the changes to `_nbextension_paths` in `__init__.py`

I copied the version from the widget cookiecutter: https://github.com/jupyter-widgets/widget-cookiecutter/blob/969471848da747d5aa562b8ab00988e3117e4ccf/%7B%7Bcookiecutter.github_project_name%7D%7D/%7B%7Bcookiecutter.python_package_name%7D%7D/__init__.py#L25

which uses:
```
    return [{
        'section': 'notebook',
        'src': 'nbextension',
        'dest': '{{ cookiecutter.npm_package_name}}',
        'require': '{{ cookiecutter.npm_package_name }}/extension'
    }]
```

but originally this cookiecutter used the python path:
https://github.com/jupyter-widgets/widget-ts-cookiecutter/blob/78716a69743e65865e89d097c2a28437b9d62820/%7B%7Bcookiecutter.github_project_name%7D%7D/%7B%7Bcookiecutter.python_package_name%7D%7D/nbextension/__init__.py#L7-L13

and since the npm package name has a slash in it `@jupyter-widgets/jupyter-widget-testwidgets` it's getting put into `share/jupyter/nbextensions/@jupyter-widgets/jupyter-widget-testwidgets` but it seems that notebook expects it to be in just `.../nbextensions/jupyter-widget-testwidgets`


Do notebooks for widgets work when there is a slash in the npm package name (or does the other cookiecutter work when there is a slash in the name)?

**Comment by ianhi** [comment]
Ahh, this is discussed in the widget-cookiecutter: https://github.com/jupyter-widgets/widget-cookiecutter/pull/80 I think the python name is the best approach

**Comment by jtpio** [comment]
> Ahh, this is discussed in the widget-cookiecutter: [jupyter-widgets/widget-cookiecutter#80](https://github.com/jupyter-widgets/widget-cookiecutter/pull/80) I think the python name is the best approach

Looks like it would make sense for the nbextension yes.

File: {{cookiecutter.github_project_name}}/{{cookiecutter.python_package_name}}/__init__.py [hunk 1]
```diff
@@ -7,4 +7,43 @@
...
+        from `src` directory into <jupyter path>/nbextensions/<dest> directory
+        during widget installation
+    require: Path to importable AMD Javascript module inside the
+        <jupyter path>/nbextensions/<dest> directory
+    """
+    return [{
+        'section': 'notebook',
+        'src': 'nbextension',
+        'dest': '{{ cookiecutter.npm_package_name}}',
+        'require': '{{ cookiecutter.npm_package_name }}/extension'
```
Comment (jtpio) [inline {{cookiecutter.github_project_name}}/{{cookiecutter.python_package_name}}/__init__.py:48]:
```suggestion
        'dest': '{{ cookiecutter.npm_package_name}}',
        'require': '{{ cookiecutter.npm_package_name }}/extension'
```
```suggestion
        'dest': '{{ cookiecutter.python_package_name }}',
        'require': '{{ cookiecutter.python_package_name }}/extension'
```

File: {{cookiecutter.github_project_name}}/setup.py [hunk 2]
```diff
@@ -6,55 +6,58 @@
...
     ]
 }
 
+
 data_files_spec = [
-    ('share/jupyter/nbextensions/{{ cookiecutter.python_package_name}}',
-        nb_path, '*.js*'),
-    ('share/jupyter/lab/extensions', lab_path, '*.tgz'),
-    ('etc/jupyter/nbconfig/notebook.d' , HERE, '{{ cookiecutter.python_package_name}}.json')
+    ('share/jupyter/nbextensions/{{ cookiecutter.npm_package_name }}', '{{ cookiecutter.python_package_name }}/nbextension', '**'),
```
Comment (jtpio) [inline {{cookiecutter.github_project_name}}/setup.py:50]:
```suggestion
    ('share/jupyter/nbextensions/{{ cookiecutter.python_package_name }}', '{{ cookiecutter.python_package_name }}/nbextension', '**'),
```

**Comment by jtpio** [comment]
@ianhi looks like CI is now passing after switching to `python_name` for the `nbextension`.

**Comment by ianhi** [comment]
> @ianhi looks like CI is now passing after switching to python_name for the nbextension.

Nice! thanks for fixing that. I've got to do lots of lab stuff today, but I'll try to test out a generate extension in both notebook and lab this evening just to make sure it really does show up properly.

**Comment by jtpio** [comment]
> I'll try to test out a generate extension in both notebook and lab this evening just to make sure it really does show up properly.

Thanks!
This is where some basic karma / jest tests help. We could also try to introduce end to end tests at some point too.

I just tried it locally with the following and it seems to be working fine :+1: 

```bash
cookiecutter https://github.com/ianhi/widget-ts-cookiecutter --checkout jlab-3
cd widgetexample
python -m pip install -e .

# check the extensions are installed
jupyter nbextension list
jupyter labextension list

# test in lab and notebook
jupyter lab example/introduction.ipynb
jupyter notebook example/introduction.ipynb
```

**Comment by ianhi** [comment]
> I just tried it locally with the following and it seems to be working fine +1

Awesome - I see the same!

So I think this is good to merge unless there are any complaints?

File: README.md [hunk 3]
```diff
@@ -54,8 +54,8 @@ When developing your extensions, you need to manually enable your extensions wit
 notebook / lab frontend. For lab, this is done by the command:
 
 ```
-jupyter labextension install @jupyter-widgets/jupyterlab-manager --no-build
-jupyter labextension install .
+jupyter labextension develop --overwrite .
+npm run build
```
Comment (jtpio) [inline README.md:58]: Something that was discussed in https://github.com/jupyter-widgets/widget-cookiecutter/pull/72.

Maybe we'll want to recommend using `yarn`, so it stays consistent with the `jlpm` and doesn't mix both `yarn.lock` and `package-lock.json` files?

Comment (ianhi) [reply README.md:58]: Sure - should we also change all the scripts in the package.json then?

Comment (jtpio) [reply README.md:58]: Probably yes. I think it's used in the widget cookiecutter here: 

https://github.com/jupyter-widgets/widget-cookiecutter/blob/969471848da747d5aa562b8ab00988e3117e4ccf/%7B%7Bcookiecutter.github_project_name%7D%7D/js/package.json#L24

**Review by jtpio — APPROVED** [review]
Thanks!

**Pull request merged**